examples/euler.cpp

```diff
@@ -0,0 +1,157 @@
+  cout << endl;
+
+  start = high_resolution_clock::now();
+
+  KeyGenerator kgen = newKeyGenerator(params);
+
+  SecretKey sk = genSecretKey(kgen);
+
+  RelinearizationKey rlk = genRelinKey(kgen, sk);
+
+  Encryptor encryptor = newEncryptor(params, sk);
+
+  Decryptor decryptor = newDecryptor(params, sk);
+
+  Encoder encoder = newEncoder(params);
+
+  Evaluator evaluator = newEvaluator(params, makeEvaluationKey(rlk));
+
+  cout << "Done in " << duration_cast<seconds>(high_resolution_clock::now() - start).count() << " seconds." << endl;
+
+  cout << endl;
+  cout << "CKKS parameters: logN = " << logN(params) << ", logSlots = " << logSlots(params) << ", logQP = " << logQP(params) << ", levels = " << maxLevel(params) + 1 << ", scale= " << scale(params) << ", sigma = " << sigma(params) << endl;
+
+  cout << endl;
+  cout << "=========================================" << endl;
+  cout << "           PLAINTEXT CREATION            " << endl;
+  cout << "=========================================" << endl;
+  cout << endl;
+
+  start = high_resolution_clock::now();
+
+  double r = 16;
+
+  double pi = 3.141592653589793;
+
+  uint64_t slots = numSlots(params);
+
+  vector<double> values(slots, 2*pi);
+
+  Plaintext plaintext = newPlaintext(params, maxLevel(params));
+  setScale(plaintext, scale(plaintext) / r);
+  encode(encoder, values, plaintext);
+
+  cout << "Done in " << duration_cast<seconds>(high_resolution_clock::now() - start).count() << " seconds." << endl;
+
+  cout << endl;
+  cout << "=========================================" << endl;
+  cout << "              ENCRYPTION                 " << endl;
+  cout << "=========================================" << endl;
+  cout << endl;
+
+  start = high_resolution_clock::now();
+
+  Ciphertext ciphertext = encryptNew(encryptor, plaintext);
+
+  cout << "Done in " << duration_cast<seconds>(high_resolution_clock::now() - start).count() << " seconds." << endl;
+
+  printDebug(params, ciphertext, values, decryptor, encoder);
+
+  cout << endl;
+  cout << "===============================================" << endl;
+  cout << "        EVALUATION OF i*x on " << slots << " values" << endl;
+  cout << "===============================================" << endl;
+  cout << endl;
+
+  start = high_resolution_clock::now();
+
+  multByConst(evaluator, ciphertext, 2, ciphertext);
+
+  cout << "Done in " << duration_cast<seconds>(high_resolution_clock::now() - start).count() << " seconds." << endl;
+
+  for (int i=0; i < values.size(); i++) {
+    values[i] *= 2;
+  }
+
+  printDebug(params, ciphertext, values, decryptor, encoder);
+
+  cout << endl;
+  cout << "===============================================" << endl;
+  cout << "       EVALUATION of x/r on " << slots << " values" << endl;
+  cout << "===============================================" << endl;
+  cout << endl;
+
+  start = high_resolution_clock::now();
+
+  setScale(ciphertext, scale(ciphertext) * r);
```

Comment:
Could you help me understand what's going on here?
- Why would you need to explicitly set the ciphertext scale? Isn't that tracked internally by Lattigo?
- If I understand this correctly, the correct ciphertext scale is ~`r^2` if the original plaintext had scale `r`, since multiplying by a scalar squares the scale.
- Below you divide the plaintext by `r`, which definitely seems fishy. Scale shouldn't affect the plaintext at all, and you shouldn't need to account for ciphertext scale on decryption.

Reply:
This is a mirroring of the CKKS/Euler example in the Lattigo lib https://github.com/tuneinsight/lattigo/blob/5b707142db0fc16acad96c1e46e7a9d68fb5b014/examples/ckks/euler/main.go

The code multiplies the scale of the ciphertext by `r`, this is equivalent to dividing the content of the ciphertext by `r` since at decoding time the decoder divides the plaintext polynomial by the scale.